tools/wrapper/retrace.c

```diff
@@ -0,0 +1,62 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+/*
+ * print usage
+ */
+
+static void usage(const char *prog_name)
+{
+	fprintf(stderr, "Usage: %s -l <path of libretrace.so> <command line>\n",
+			prog_name);
+	exit(1);
+}
+
+/*
+ * fork process
+ */
+
+static void fork_process(char **argv)
+{
+	pid_t pid;
+	int w, status;
+
+	/* fork new process */
+	pid = fork();
+	if (pid == 0) {
+		/* set environment for preload */
+		setenv("LD_PRELOAD", argv[2], 1);
+		exit(execvp(argv[3], &argv[3]));
+	} else if (pid < 0) {
+		fprintf(stderr, "Failed to fork new process(err:%d)\n", errno);
+		exit(1);
+	}
+
+	/* wait until process has terminated */
+	w = waitpid(pid, &status, 0);
+	if (w < 0)
+		fprintf(stderr, "waitpid() error for command\n");
+}
+
+/*
+ * main function
+ */
+
+int main(int argc, char *argv[])
+{
+	/* check arguments */
+	if (argc < 4)
+		usage(argv[0]);
+
+	if (strcmp(argv[1], "-l") != 0)
+		usage(argv[0]);
+
+	fork_process(argv);
```

Comment:
`argv + 3` to make `fork_process` more generic ... although will be `argv + optind` once you implement option parsing with `getopt`